File: react/src/classes/EventDispatcher.ts
export class EventDispatcher<T = any> {
  private listeners: Record<string, ((payload?: T) => void)[]> = {};

  addEventListener(eventName: string, cb: (payload?: T) => void) {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(cb);
  }

  dispatchEvent(eventName: string, payload?: T) {
    if (!this.listeners[eventName]) return;
    this.listeners[eventName].forEach((cb) => cb(payload));
  }
}

File: src/inc/EventDispatcher.cpp
#include "EventDispatcher.h"

void EventDispatcher::addEventListener(const std::string &eventName, Callback cb)
{
    listeners[eventName].push_back(cb);
}

void EventDispatcher::dispatchEvent(const std::string &eventName, void *payload)
{
    if (listeners.find(eventName) == listeners.end())
        return;
    for (const auto &cb : listeners[eventName])
    {
        cb(payload);
    }
}
